catalyst_voices/lib/widgets/rich_text/voices_rich_text.dart

```diff
@@ -0,0 +1,259 @@
+import 'package:flutter_quill_extensions/flutter_quill_extensions.dart';
+
+/// A component for rich text writing
+/// using Quill under the hood
+/// https://pub.dev/packages/flutter_quill
+class VoicesRichText extends StatefulWidget {
+  const VoicesRichText({
+    super.key,
+    this.document,
+    this.onSave,
+    this.charsLimit,
+  });
+
+  final Document? document;
+  final ValueChanged<Document>? onSave;
+  final int? charsLimit;
+
+  @override
+  State<VoicesRichText> createState() => _VoicesRichTextState();
+}
+
+class _VoicesRichTextState extends State<VoicesRichText> {
+  final QuillController _controller = QuillController.basic();
+  int _documentLength = 0;
+  bool _editMode = false;
+  final FocusNode _focusNode = FocusNode();
+
+  @override
+  void initState() {
+    super.initState();
+    if (widget.document != null) _controller.document = widget.document!;
+    _controller.document.changes.listen(_onDocumentChange);
+    _documentLength = _controller.document.length;
+  }
+
+  void _onDocumentChange(DocChange docChange) {
+    final documentLength = _controller.document.length;
+    setState(() {
+      _documentLength = documentLength;
+    });
+    final limit = widget.charsLimit;
+
+    if (limit == null) return;
+
+    if (documentLength > limit) {
+      final latestIndex = limit - 1;
+      _controller.replaceText(
+        latestIndex,
+        documentLength - limit,
+        '',
+        TextSelection.collapsed(offset: latestIndex),
+      );
+    }
+  }
+
+  @override
+  void dispose() {
+    _controller.dispose();
+    _focusNode.dispose();
+    super.dispose();
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    final theme = Theme.of(context);
+
+    return ExcludeFocus(
+      excluding: !_editMode,
+      child: Column(
+        children: [
+          Padding(
+            padding: const EdgeInsets.only(
+              left: 16,
+              top: 20,
+              bottom: 20,
+            ),
+            child: Row(
+              children: [
+                Text(
+                  'Rich text',
+                  style: theme.textTheme.titleMedium,
+                ),
+                Expanded(
+                  child: Align(
+                    alignment: Alignment.centerRight,
+                    child: Padding(
+                      padding: const EdgeInsets.symmetric(horizontal: 8.0),
+                      child: TextButton(
+                        onPressed: () {
+                          setState(() {
+                            _editMode = !_editMode;
+                          });
+                        },
+                        child: Text(
+                          _editMode ? 'Cancel' : 'Edit', //TODO localize
+                          style: theme.textTheme.labelSmall,
+                        ),
+                      ),
+                    ),
+                  ),
+                )
+              ],
+            ),
+          ),
```

Comment:
Are you sure it does makes sense to have this header with `Rich text` / `Cancel/Edit` buttons as part of `VoicesRichText`? 

I would say it should be out of this widget because we'll probably wont need them everywhere where we use RichText.

Reply:
I double checked with Niels, he is sure that they will be always used together.